### PR TITLE
add a TypedAST node that holds a type func ID

### PR DESCRIPTION
**Pull request by SebastianMestre**
An idea: compile time is just another moment for evaluation, just that most expressions evaluate to themselves, and some evaluate to monotypes or type functions.

Example: `struct { x : int }` evaluates into a type function

### CODE REVIEW

**Comment by SebastianMestre** [comment]
whew... this was exhausting.

**Comment by SebastianMestre** [comment]
RATIONALE:

It makes sense to think about a compile-time declarative language that deals in types and expressions. Using this language, we build up the type functions and monotypes that are used in the final program.

In the implementation of this language, TypedAST::TypedAST is analogous to Interpreter::Value.

**Comment by lushoBarlett** [comment]
Makes sense, although I will read this tomorrow because its suicide to do it now.

File: src/typechecker.cpp [hunk 1]
```diff
@@ -176,8 +176,16 @@ void TypeChecker::declare_builtin(std::string const& name, MetaTypeId meta_type,
...
+	// BIG HACK:
+	// if we are declaring a typefunc, 'poly_type' is actually a TypeFunctionId
+	if (meta_type == meta_typefunc()) {
+		auto handle = std::make_unique<TypedAST::TypeFunctionHandle>();
+		handle->m_value = poly_type;
+		decl->m_value = std::move(handle);
+	} else {
+		decl->m_decl_type = poly_type;
+		decl->m_is_polymorphic = true;
+	}
```
Comment (lushoBarlett) [inline src/typechecker.cpp:188]: Not a fan of this, will look into ways of cleaning this up

Comment (SebastianMestre) [reply src/typechecker.cpp:188]: yeah it's disgusting

**Review by lushoBarlett — APPROVED** [review]
So, yea, most of the code added was just tree traversal boilerplate, and it only does one thing for now. The TypeFunctionHandle is a straightforward approach, similar to how identifier points to the declaration.

I must say that I will attempt to read all of the code and look for ways to reduce the amount of lines in it. I get very confused and it takes way too long for me to review this stuff.

I think this is correct.

**Comment by SebastianMestre** [comment]
The tricky bits are in `type_func_from_ast`, `ct_eval(DeclarationList)`, and, as you noted, `Typechecker::declare_builtin`

**Pull request merged**